r/tests/testthat/test-dplyr.R

```diff
@@ -133,6 +139,76 @@ test_that("filtering with expression", {
   )
 })
 
+test_that("filtering with arithmetic", {
+  expect_dplyr_equal(
+    input %>%
+      filter(dbl + 1 > 3) %>%
+      select(string = chr, int, dbl) %>%
+      collect(),
+    tbl
+  )
+
+  expect_dplyr_equal(
+    input %>%
+      filter(dbl / 2 > 3) %>%
+      select(string = chr, int, dbl) %>%
+      collect(),
+    tbl
+  )
+
+  expect_dplyr_equal(
+    input %>%
+      filter(dbl / 2L > 3) %>%
+      select(string = chr, int, dbl) %>%
+      collect(),
+    tbl
+  )
+
+  expect_dplyr_equal(
+    input %>%
+      filter(int / 2 > 3) %>%
+      select(string = chr, int, dbl) %>%
+      collect(),
+    tbl
+  )
+
+  expect_dplyr_equal(
+    input %>%
+      filter(int / 2L > 3) %>%
+      select(string = chr, int, dbl) %>%
+      collect(),
+    tbl
+  )
+
+  skip("autocasting should happen in compute kernels; R workaround fails on this ARROW-11078")
```

Comment:
Does this fail for the same reason that the comparisons below fail? `int / 2L` is an integer but the `array_expression` doesn't know the resulting type, so it can't cast `3` to it?

How does it fail? Gracefully or not?

We could work around this either by tracking/guessing the resulting type and sticking it in the array_expression object or by deferring the autocasting until the expressions are evaluated (so we'll know the type of `int > 2L` when we evaluate `that > 3`), but it's probably not worth it now. Maybe make a jira for us to come back to?